libnavigation-core/src/main/java/com/mapbox/navigation/core/HistoryRecordingStateHandler.kt

```diff
@@ -0,0 +1,102 @@
+package com.mapbox.navigation.core
+
+import androidx.annotation.UiThread
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.core.internal.HistoryRecordingStateChangeObserver
+import com.mapbox.navigation.core.trip.session.NavigationSessionState
+import com.mapbox.navigation.core.trip.session.NavigationSessionUtils
+import com.mapbox.navigation.core.trip.session.TripSessionState
+import com.mapbox.navigation.core.trip.session.TripSessionStateObserver
+import com.mapbox.navigation.utils.internal.logI
+
+private const val TAG = "HistoryRecordingStateHandler"
+
+@UiThread
+internal class HistoryRecordingStateHandler(
+    private var currentState: NavigationSessionState
+) : TripSessionStateObserver {
+
+    private val historyRecordingStateChangeObservers =
+        mutableSetOf<HistoryRecordingStateChangeObserver>()
+
+    private var hasRoutes = false
+
+    private var isDriving = false
+
+    fun registerStateChangeObserver(observer: HistoryRecordingStateChangeObserver) {
+        historyRecordingStateChangeObservers.add(observer)
+    }
+
+    fun unregisterStateChangeObserver(observer: HistoryRecordingStateChangeObserver) {
+        historyRecordingStateChangeObservers.remove(observer)
+    }
+
+    fun unregisterAllStateChangeObservers() {
+        historyRecordingStateChangeObservers.clear()
+    }
+
+    fun setRoutes(routes: List<NavigationRoute>) {
+        logI("setRoutes: $routes", TAG)
+        if (updateHasRoutes(routes.isNotEmpty())) {
+            updateStateAndNotifyObservers { observer, state ->
+                observer.onShouldStopRecording(state)
+            }
+        }
+    }
+
+    fun setRoutesFailed() {
+        logI("setRoutes failed", TAG)
+        if (updateHasRoutes(false)) {
```

Comment:
> if (updateHasRoutes(false)) {

I don't think this condition should always resolve to false 🤔 

What happens when
1. Set routes to not emtpy and it is successful
2. New route request fails and the new routes are ignored

There will still be active routes, but this state handler will think there are no routes

Reply:
Makes sense, thanks for noticing. I should add a test for this scenario as well.